### PR TITLE
Represent non-stringified JSON request body as an [object Object] string

### PR DESCRIPTION
**Pull request by kettanaito**
## Changes

Ensures `request._bodyInit` (the internal representation of a request body) is always set to a `[object Object]` string representation of a body in case it has a non-stringified JSON body and the `Content-Type` header stating an expected JSON body.

## Why

- See #856

### CODE REVIEW

**Comment by kettanaito** [comment]
I'd like to ask for help with adding a test for this. Following the existing tests as examples, I've written this test:

```js
test('converts a non-stringified json body to [object Object] string', function() {
  return fetch('/request', {
    method: 'post',
    headers: {
      'content-type': 'application/json'
    },
    body: {stringified: false}
  })
    .then(function(response) {
      return response.json()
    })
    .then(function(request) {
      assert.equal(request.data, '[object Object]')
    })
})
```

However, `request.data` is a public request body that gets correctly set to `[object Object]`. The test doesn't fail when the fix from this pull request is reverted. I can only assume that `request.data` doesn't equal the request body provided to `xhr.send()` at the end of a request:

https://github.com/github/fetch/blob/a8aa427de0ed808ff26c0e3eb2e59c122c44488a/fetch.js#L594

Assertions on `request._bodyInit` fail, that doesn't seem to be public property (dangle implies).

File: fetch.js [hunk 1]
```diff
@@ -246,14 +246,20 @@ function Body() {
...
+
+    if (!contentType) {
       if (typeof body === 'string') {
         this.headers.set('content-type', 'text/plain;charset=UTF-8')
       } else if (this._bodyBlob && this._bodyBlob.type) {
         this.headers.set('content-type', this._bodyBlob.type)
       } else if (support.searchParams && URLSearchParams.prototype.isPrototypeOf(body)) {
         this.headers.set('content-type', 'application/x-www-form-urlencoded;charset=UTF-8')
       }
+    } else if (contentType.includes('json') && typeof this._bodyInit !== 'string') {
```
Comment (kettanaito) [inline fetch.js:259]: This isn't a reliable JSON request body content-type header check. Would you recommend to provide `application/json; charset=UTF-8` instead? 

I'm thinking that this behavior should also affect _any_ JSON-like content-type header (i.e. `application/hal+json`).

**Comment by lektor2034** [comment]
thk sir

File: fetch.js [hunk 2]
```diff
@@ -246,14 +246,20 @@ function Body() {
       this._bodyText = body = Object.prototype.toString.call(body)
```
Comment (kettanaito) [inline fetch.js:246]: Alternatively, we can set `this._bodyInit` to equal `this._bodyText` in this closure. I'm not sure about the implications of doing this, so would appreciate a feedback.

Comment (whiteslack) [reply fetch.js:246]: whiteslack

**Review by whiteslack — APPROVED** [review]
whiteslack

**Review by Toppole69 — APPROVED** [review]

**Review by Toppole69 — APPROVED** [review]

**Review by SonyViao — APPROVED** [review]

**Review by SonyViao — APPROVED** [review]

**Review by SonyViao — APPROVED** [review]

File: fetch.js [hunk 3]
```diff
@@ -246,14 +246,20 @@ function Body() {
...
     }
 
-    if (!this.headers.get('content-type')) {
```
Comment (marcin543) [inline fetch.js:249]: - [ ]

File: fetch.js [hunk 4]
```diff
@@ -246,14 +246,20 @@ function Body() {
...
+    const contentType = this.headers.get('content-type')
```
Comment (marcin543) [inline fetch.js:249]: - [ ]

File: fetch.js (hunk 1 above)
Comment (marcin543) [reply fetch.js:259]: - [ ]

File: fetch.js [hunk 5]
```diff
@@ -246,14 +246,20 @@ function Body() {
...
       } else if (this._bodyBlob && this._bodyBlob.type) {
         this.headers.set('content-type', this._bodyBlob.type)
       } else if (support.searchParams && URLSearchParams.prototype.isPrototypeOf(body)) {
         this.headers.set('content-type', 'application/x-www-form-urlencoded;charset=UTF-8')
       }
+    } else if (contentType.includes('json') && typeof this._bodyInit !== 'string') {
+      // Always pass a text representation of a non-stringified JSON body
+      // to `XMLHttpRequest.send` to retain a compatible behavior with the browser.
+      this._bodyInit = this._bodyText
     }
```
Comment (marcin543) [inline fetch.js:263]: - [ ]

**Review by marcin543 — APPROVED** [review]

**Review by skkienn27angela — APPROVED** [review]
>

**Comment by skkienn27angela** [comment]
``

**Review by JakeChampion — APPROVED** [review]

**Comment by JakeChampion** [comment]
This is released in v3.6.0 -- https://github.com/github/fetch/releases/tag/v3.6.0

**Review by NeverLiez — COMMENTED** [review]
B

**Review by Apple1D — COMMENTED** [review]
- [ ]

**Pull request merged**